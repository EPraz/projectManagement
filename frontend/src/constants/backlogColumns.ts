// Actualiza tu interfaz para incluir una propiedad de orden
export interface BacklogColumnsProps {
  key: string;
  label: string;
  order: number; // Añadimos una propiedad de orden explícita
  width?: string; // Opcional: ancho personalizado
}

// Actualiza tus columnas con la propiedad de orden
export const backlogColumns: BacklogColumnsProps[] = [
  { key: "order", label: "Order", order: 1, width: "10%" },
  { key: "id", label: "ID", order: 2, width: "10%" },
  { key: "state", label: "State", order: 3, width: "15%" },
  { key: "title", label: "Title", order: 4 },
  { key: "type", label: "Type", order: 5 },
  { key: "priority", label: "Priority", order: 6, width: "10%" },
  { key: "assignedTo", label: "Assigned To", order: 7 },
  { key: "sprint", label: "Sprint", order: 8 },
  { key: "dueDate", label: "Due Date", order: 9 },
  { key: "estimatedHours", label: "Est. Hours", order: 10, width: "8%" },
  { key: "remainingHours", label: "Rem. Hours", order: 11, width: "8%" },
  { key: "isBlocked", label: "Blocking", order: 12 },
  { key: "tags", label: "Tags", order: 13 },
  { key: "actions", label: "Actions", order: 14, width: "5%" },
];

// Columnas inicialmente visibles
export const initialBacklogColumns: string[] = [
  "order",
  "id",
  "state",
  "title",
  "priority",
  "tags",
  "actions",
];
