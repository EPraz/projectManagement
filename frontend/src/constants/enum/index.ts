export * from "./Role";
export * from "./TicketPriority";
export * from "./TicketType";
