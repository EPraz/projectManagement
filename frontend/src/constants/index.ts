export * from "./taskStatus";
export * from "./ticketStatus";
export * from "./enum";
export * from "./menuItems";
export * from "./tabPanels";
export * from "./backlogColumns";
export * from "./apiConstant";
