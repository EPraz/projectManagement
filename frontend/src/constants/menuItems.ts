import DashboardIcon from "@mui/icons-material/Dashboard";
import ViewKanbanIcon from "@mui/icons-material/ViewKanban";
import AssessmentIcon from "@mui/icons-material/Assessment";
import CategoryIcon from "@mui/icons-material/Category";
import PeopleIcon from "@mui/icons-material/People";
import SettingsIcon from "@mui/icons-material/Settings";

export interface MenuItem {
  text: string;
  icon: React.ElementType;
  path: string;
  disabled?: boolean;
}

export const MENU_ITEMS: MenuItem[] = [
  {
    text: "Overview",
    icon: DashboardIcon,
    path: "overview",
  },
  {
    text: "Board",
    icon: ViewKanbanIcon,
    path: "board",
  },
  {
    text: "Retrospective",
    icon: AssessmentIcon,
    path: "retrospective",
  },
  {
    text: "Epics & Features",
    icon: CategoryIcon,
    path: "epics",
    disabled: true,
  },

  {
    text: "Team Members",
    icon: PeopleIcon,
    path: "teammembers",
  },
  {
    text: "Settings",
    icon: SettingsIcon,
    path: "settings",
  },
];

export interface MainTab {
  text: string;
  path: string;
}

export const MAIN_TABS: MainTab[] = [
  {
    text: "Overview",
    path: "overview",
  },
  {
    text: "Board",
    path: "board",
  },
  {
    text: "Members",
    path: "members",
  },
  {
    text: "Files",
    path: "files",
  },
  {
    text: "Timeline",
    path: "timeline",
  },
  {
    text: "Reports",
    path: "reports",
  },
];

export interface Project {
  name: string;
  count: number;
  color: string;
}

export const PROJECTS: Project[] = [
  { name: "ShieldSafe", count: 5, color: "#1E8E3E" },
  { name: "Quantum", count: 2, color: "#1A73E8" },
  { name: "AeroSquare", count: 7, color: "#F9AB00" },
];

export const cardVariants = {
  hidden: { opacity: 0, y: 20 },
  visible: { opacity: 1, y: 0 },
};
