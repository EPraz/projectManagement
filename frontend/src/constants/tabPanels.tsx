import ViewKanbanIcon from "@mui/icons-material/ViewKanban";
import FormatListBulletedIcon from "@mui/icons-material/FormatListBulleted";
import GroupIcon from "@mui/icons-material/Group";
import FlagIcon from "@mui/icons-material/Flag";

import { Taskboard } from "../pages";
import BacklogBoard from "../pages/board/backlogBoard/BacklogBoard";
import CapacityBoard from "../pages/board/capacityBoard/CapacityBoard";
import GoalBoard from "../pages/board/goalBoard/GoalBoard";

interface TabPanel {
  label: string;
  icon: JSX.Element;
  component: () => JSX.Element;
}

export const TabPanels: TabPanel[] = [
  { label: "Board", icon: <ViewKanbanIcon />, component: () => <Taskboard /> },
  {
    label: "Backlog",
    icon: <FormatListBulletedIcon />,
    component: () => <BacklogBoard />,
  },
  {
    label: "Capacity",
    icon: <GroupIcon />,
    component: () => <CapacityBoard />,
  },
  { label: "Goals", icon: <FlagIcon />, component: () => <GoalBoard /> },
];
