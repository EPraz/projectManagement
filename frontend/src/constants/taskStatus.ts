export const TASK_STATUSES = [
  // to be deleted
  "To Do",
  "In Progress",
  "Blocked",
  "Review",
  "Testing",
  "Done",
];
