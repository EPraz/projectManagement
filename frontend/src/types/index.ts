export * from "./contextProps";
export * from "./models";
export * from "./materialUiExtensions";
export * from "./components";
export * from "./FilterProps";
