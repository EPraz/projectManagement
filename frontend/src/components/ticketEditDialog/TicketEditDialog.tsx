import React, { useState, useEffect, useCallback } from "react";
import {
  DialogContent,
  TextField,
  Button,
  Box,
  Typography,
  IconButton,
  FormControl,
  InputLabel,
  Select,
  MenuItem,
  Avatar,
  Tab,
  Chip,
  SelectChangeEvent,
} from "@mui/material";
import {
  Close as CloseIcon,
  Save as SaveIcon,
  Delete as DeleteIcon,
  Block as BlockIcon,
} from "@mui/icons-material";
import type { Ticket, TicketEditDialogProps } from "../../types";
import { TicketPriority, TicketType } from "../../constants";
import { getPriorityColor, getTypeIcon, omitProps } from "../../helpers";
import { useAuth, useSprint } from "../../context";
import {
  Header,
  LeftPanel,
  MainContent,
  PriorityChip,
  RightPanel,
  Section,
  SectionTitle,
  StyledDialog,
  StyledTabs,
  TimeTrackingGrid,
} from "./TicketEditDialog.styles";

const TicketEditDialog: React.FC<TicketEditDialogProps> = ({
  open,
  onClose,
  ticket,
  onSave,
  onDelete,
  statuses,
  users,
  sprints,
  features,
  projects,
  tickets,
  disabled = false,
}) => {
  const { sprint } = useSprint();
  const [editedTicket, setEditedTicket] = useState<Partial<Ticket> | null>(
    null
  );
  const [errors, setErrors] = useState<Record<string, string>>({});
  const [activeTab, setActiveTab] = useState(0);
  const { user: currentUser } = useAuth();

  useEffect(() => {
    if (ticket) {
      setEditedTicket({ ...ticket });
    } else {
      setEditedTicket(null);
    }
    setErrors({});
    setActiveTab(0);
  }, [ticket, open]);

  const handleChange = (
    e:
      | React.ChangeEvent<HTMLInputElement | { name?: string; value: unknown }>
      | SelectChangeEvent<string>
  ) => {
    const { name, value } = e.target;
    if (!editedTicket || !name) return;

    setEditedTicket({
      ...editedTicket,
      [name]: value,
    });

    // Clear error when field is edited
    if (errors[name]) {
      setErrors({
        ...errors,
        [name]: "",
      });
    }
  };

  const handleBlockingTicketChange = (e: SelectChangeEvent<string>) => {
    // Si el valor es "", lo convertimos a null en lugar de undefined.
    const selectedTicketId =
      e.target.value === "" ? null : Number(e.target.value);
    setEditedTicket({
      ...editedTicket,
      blockedBy: selectedTicketId,
    });
  };

  const handleStatusChange = (e: SelectChangeEvent<string>) => {
    const { name } = e.target;
    if (!editedTicket || !name || editedTicket?.isBlocked) return;
    const newStatusId = e.target.value as string;
    setEditedTicket({
      ...editedTicket,
      statusId: newStatusId,
    });
  };

  const handleBlockedChange = () => {
    if (disabled) return;
    const newBlocked = !editedTicket?.isBlocked;
    setEditedTicket({
      ...editedTicket,
      isBlocked: newBlocked,
      statusId: newBlocked
        ? statuses?.find((x) => x.name === "BLOCKED")?.id ||
          editedTicket?.statusId
        : "",
      blockedBy: newBlocked ? editedTicket?.blockedBy : null,
    });
  };

  // const handleDateChange = (date: Date | null, fieldName: string) => {
  //   if (!editedTicket) return;

  //   setEditedTicket({
  //     ...editedTicket,
  //     [fieldName]: date ? date.toISOString() : undefined,
  //   });
  // };

  const handleNumberChange = (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    fieldName: string
  ) => {
    if (!editedTicket) return;

    const value = e.target.value === "" ? undefined : Number(e.target.value);

    setEditedTicket({
      ...editedTicket,
      [fieldName]: value,
    });
  };

  const handleTagsChange = (
    event: React.SyntheticEvent,
    newValue: string[]
  ) => {
    if (!editedTicket) return;

    setEditedTicket({
      ...editedTicket,
      tags: newValue,
    });
  };

  const validateForm = (): boolean => {
    const newErrors: Record<string, string> = {};

    if (!editedTicket?.title) {
      newErrors.title = "Title is required";
    }

    if (!editedTicket?.statusId) {
      newErrors.statusId = "Status is required";
    }

    if (!editedTicket?.priority) {
      newErrors.priority = "Priority is required";
    }

    if (!editedTicket?.type) {
      newErrors.type = "Type is required";
    }

    setErrors(newErrors);
    return Object.keys(newErrors).length === 0;
  };

  const handleSave = useCallback(() => {
    if (!editedTicket || !validateForm()) return;

    const dataToSend = omitProps(editedTicket, [
      "createdBy",
      "createdAt",
      "updatedAt",
      "featureId",
      "projectId",
      "project",
      "tasks",
      "status",
      "blockingTicket",
      "blockedTickets",
      "assignedUser",
      "order",
      "_count",
      "sprint",
      "feature",
      "pairProgrammingUsers",
    ]);

    onSave({
      ...dataToSend,
      updatedBy: currentUser?.email,
      sprintId: sprint?.id,
    });
    onClose();
  }, [editedTicket, currentUser, onSave, onClose]);

  const handleDelete = useCallback(() => {
    if (!editedTicket || !onDelete) return;
    onDelete(editedTicket);
    onClose();
  }, [editedTicket, onDelete, onClose]);

  if (!editedTicket) return null;

  return (
    <StyledDialog open={open} onClose={onClose} fullWidth maxWidth="lg">
      <Header>
        <Box display="flex" justifyContent="space-between" alignItems="center">
          <Box display="flex" alignItems="center" gap={1}>
            <Typography variant="h6">
              {ticket?.id ? "Ticket Details" : "New Ticket"}
            </Typography>
            {editedTicket.priority && (
              <PriorityChip
                size="small"
                label={editedTicket.priority}
                priorityColor={getPriorityColor(editedTicket.priority)}
              />
            )}
          </Box>
          <IconButton onClick={onClose} size="small">
            <CloseIcon />
          </IconButton>
        </Box>
      </Header>

      <DialogContent sx={{ p: 0 }}>
        <MainContent>
          <LeftPanel>
            <StyledTabs
              value={activeTab}
              onChange={(_, newValue) => setActiveTab(newValue)}
              variant="scrollable"
              scrollButtons="auto"
            >
              <Tab label="Details" />
              <Tab label="Planning" />
              <Tab label="Additional Info" />
            </StyledTabs>

            {activeTab === 0 && (
              <>
                <Section>
                  <SectionTitle>Ticket Information</SectionTitle>
                  <TextField
                    name="title"
                    placeholder="Ticket title"
                    value={editedTicket.title || ""}
                    onChange={handleChange}
                    fullWidth
                    variant="outlined"
                    size="small"
                    error={!!errors.title}
                    helperText={errors.title}
                    required
                    disabled={disabled}
                    sx={{ mb: 2 }}
                  />

                  <Box display="flex" gap={2} mb={2}>
                    <FormControl
                      fullWidth
                      size="small"
                      error={!!errors.statusId}
                    >
                      <InputLabel>Status</InputLabel>
                      <Select
                        name="statusId"
                        value={editedTicket.statusId || ""}
                        onChange={handleStatusChange}
                        label="Status"
                        required
                        disabled={disabled || editedTicket.isBlocked}
                      >
                        {statuses &&
                          statuses.map((status) => (
                            <MenuItem key={status.id} value={status.id}>
                              <Box display="flex" alignItems="center" gap={1}>
                                <Box
                                  sx={{
                                    width: 12,
                                    height: 12,
                                    borderRadius: "50%",
                                    backgroundColor: status.color || "#ccc",
                                  }}
                                />
                                {status.name}
                              </Box>
                            </MenuItem>
                          ))}
                      </Select>
                      {errors.statusId && (
                        <Typography variant="caption" color="error">
                          {errors.statusId}
                        </Typography>
                      )}
                    </FormControl>

                    <FormControl fullWidth size="small" error={!!errors.type}>
                      <InputLabel>Type</InputLabel>
                      <Select
                        name="type"
                        value={editedTicket.type || ""}
                        onChange={handleChange}
                        label="Type"
                        required
                        disabled={disabled}
                      >
                        {Object.values(TicketType).map((type) => (
                          <MenuItem key={type} value={type}>
                            <Box display="flex" alignItems="center" gap={1}>
                              {getTypeIcon(type)}
                              {type}
                            </Box>
                          </MenuItem>
                        ))}
                      </Select>
                      {errors.type && (
                        <Typography variant="caption" color="error">
                          {errors.type}
                        </Typography>
                      )}
                    </FormControl>
                  </Box>

                  <TextField
                    name="description"
                    placeholder="Add a more detailed description..."
                    value={editedTicket.description || ""}
                    onChange={handleChange}
                    fullWidth
                    multiline
                    rows={4}
                    variant="outlined"
                    size="small"
                    disabled={disabled}
                  />
                </Section>

                <Section>
                  <SectionTitle>Blocking Status</SectionTitle>
                  <Box display="flex" alignItems="center" gap={2}>
                    <Chip
                      icon={<BlockIcon />}
                      label="Blocked"
                      color={editedTicket.isBlocked ? "error" : "default"}
                      variant={editedTicket.isBlocked ? "filled" : "outlined"}
                      onClick={handleBlockedChange}
                      disabled={disabled}
                    />

                    {editedTicket.isBlocked && (
                      <FormControl fullWidth size="small">
                        <InputLabel>Blocked By Ticket</InputLabel>
                        <Select
                          name="blockedBy"
                          value={editedTicket?.blockedBy?.toString() || ""}
                          onChange={handleBlockingTicketChange}
                          label="Blocked By Ticket"
                          disabled={disabled}
                        >
                          <MenuItem value="" disabled>
                            <em>Select a ticket</em>
                          </MenuItem>
                          {tickets
                            ?.filter((t) => t.id !== editedTicket.id)
                            ?.map((ticket) => (
                              <MenuItem
                                key={ticket.id}
                                value={ticket.id.toString()}
                              >
                                #{ticket.id} - {ticket.title}
                              </MenuItem>
                            ))}
                        </Select>
                      </FormControl>
                    )}
                  </Box>
                </Section>
              </>
            )}

            {activeTab === 1 && (
              <>
                <Section>
                  <SectionTitle>Planning</SectionTitle>
                  <Box display="flex" gap={2} mb={2}>
                    <FormControl fullWidth size="small">
                      <InputLabel>Sprint</InputLabel>
                      <Select
                        name="sprintId"
                        value={editedTicket.sprintId || ""}
                        onChange={handleChange}
                        label="Sprint"
                        disabled={disabled}
                      >
                        <MenuItem value="">
                          <em>Backlog</em>
                        </MenuItem>
                        {sprints?.map((sprint) => (
                          <MenuItem key={sprint.id} value={sprint.id}>
                            {sprint.name}
                          </MenuItem>
                        ))}
                      </Select>
                    </FormControl>

                    <FormControl fullWidth size="small">
                      <InputLabel>Feature</InputLabel>
                      <Select
                        name="featureId"
                        value={editedTicket.featureId || ""}
                        onChange={handleChange}
                        label="Feature"
                        disabled={disabled}
                      >
                        <MenuItem value="">
                          <em>None</em>
                        </MenuItem>
                        {features?.map((feature) => (
                          <MenuItem key={feature.id} value={feature.id}>
                            {feature.title}
                          </MenuItem>
                        ))}
                      </Select>
                    </FormControl>
                  </Box>

                  <Box display="flex" gap={2} mb={2}>
                    <FormControl
                      fullWidth
                      size="small"
                      error={!!errors.priority}
                    >
                      <InputLabel>Priority</InputLabel>
                      <Select
                        name="priority"
                        value={editedTicket.priority || ""}
                        onChange={handleChange}
                        label="Priority"
                        required
                        disabled={disabled}
                      >
                        {Object.values(TicketPriority).map((priority) => (
                          <MenuItem key={priority} value={priority}>
                            <Box display="flex" alignItems="center" gap={1}>
                              <Box
                                sx={{
                                  width: 12,
                                  height: 12,
                                  borderRadius: "50%",
                                  backgroundColor: getPriorityColor(priority),
                                }}
                              />
                              {priority}
                            </Box>
                          </MenuItem>
                        ))}
                      </Select>
                      {errors.priority && (
                        <Typography variant="caption" color="error">
                          {errors.priority}
                        </Typography>
                      )}
                    </FormControl>
                  </Box>
                </Section>

                <Section>
                  <SectionTitle>Story Points & Estimation</SectionTitle>
                  <Box display="flex" gap={2} mb={2}>
                    <TextField
                      label="Story Points"
                      type="number"
                      value={editedTicket.storyPoints ?? ""}
                      onChange={(e) => handleNumberChange(e, "storyPoints")}
                      fullWidth
                      size="small"
                      InputProps={{
                        inputProps: { min: 0 },
                      }}
                      disabled={disabled}
                    />

                    <TextField
                      label="Estimated Hours"
                      type="number"
                      value={editedTicket.estimatedHours ?? ""}
                      onChange={(e) => handleNumberChange(e, "estimatedHours")}
                      fullWidth
                      size="small"
                      InputProps={{
                        inputProps: { min: 0, step: 0.5 },
                      }}
                      disabled={disabled}
                    />
                  </Box>

                  <Box display="flex" gap={2}>
                    <TextField
                      label="Remaining Hours"
                      type="number"
                      value={editedTicket.remainingHours ?? ""}
                      onChange={(e) => handleNumberChange(e, "remainingHours")}
                      fullWidth
                      size="small"
                      InputProps={{
                        inputProps: { min: 0, step: 0.5 },
                      }}
                      disabled={disabled}
                    />

                    <TextField
                      label="Completed Hours"
                      type="number"
                      value={editedTicket.completedHours ?? ""}
                      onChange={(e) => handleNumberChange(e, "completedHours")}
                      fullWidth
                      size="small"
                      InputProps={{
                        inputProps: { min: 0, step: 0.5 },
                      }}
                      disabled={disabled}
                    />
                  </Box>
                </Section>
              </>
            )}

            {activeTab === 2 && (
              <>
                <Section>
                  <SectionTitle>Acceptance Criteria</SectionTitle>
                  <TextField
                    name="acceptanceCriteria"
                    placeholder="Define what makes this ticket complete..."
                    value={editedTicket.acceptanceCriteria || ""}
                    onChange={handleChange}
                    fullWidth
                    multiline
                    rows={4}
                    variant="outlined"
                    size="small"
                    disabled={disabled}
                  />
                </Section>

                <Section>
                  <SectionTitle>Design Information</SectionTitle>
                  <TextField
                    name="designInformation"
                    placeholder="Add design details, links to mockups, etc..."
                    value={editedTicket.designInformation || ""}
                    onChange={handleChange}
                    fullWidth
                    multiline
                    rows={3}
                    variant="outlined"
                    size="small"
                    disabled={disabled}
                  />
                </Section>

                <Section>
                  <SectionTitle>Additional Details</SectionTitle>
                  <TextField
                    name="additionalDetails"
                    placeholder="Any other relevant information..."
                    value={editedTicket.additionalDetails || ""}
                    onChange={handleChange}
                    fullWidth
                    multiline
                    rows={3}
                    variant="outlined"
                    size="small"
                    disabled={disabled}
                  />
                </Section>

                <Section>
                  <SectionTitle>Notes & Discussion</SectionTitle>
                  <TextField
                    name="notes"
                    placeholder="Add notes..."
                    value={editedTicket.notes || ""}
                    onChange={handleChange}
                    fullWidth
                    multiline
                    rows={3}
                    variant="outlined"
                    size="small"
                    disabled={disabled}
                    sx={{ mb: 2 }}
                  />

                  <TextField
                    name="discussion"
                    placeholder="Add discussion points..."
                    value={editedTicket.discussion || ""}
                    onChange={handleChange}
                    fullWidth
                    multiline
                    rows={3}
                    variant="outlined"
                    size="small"
                    disabled={disabled}
                  />
                </Section>
              </>
            )}
          </LeftPanel>

          <RightPanel>
            <Section>
              <SectionTitle>Assignment</SectionTitle>
              <FormControl fullWidth size="small">
                <InputLabel>Assigned To</InputLabel>
                <Select
                  name="assignedTo"
                  value={editedTicket.assignedTo || ""}
                  onChange={handleChange}
                  label="Assigned To"
                  disabled={disabled}
                >
                  <MenuItem value="">
                    <em>Unassigned</em>
                  </MenuItem>
                  {users?.map((user) => (
                    <MenuItem key={user.id} value={user.id}>
                      <Box display="flex" alignItems="center" gap={1}>
                        <Avatar alt={user.name} sx={{ width: 24, height: 24 }}>
                          {user.name.charAt(0)}
                        </Avatar>
                        {user.name}
                      </Box>
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </Section>

            <Section>
              <SectionTitle>Time Tracking</SectionTitle>
              <TimeTrackingGrid>
                <TextField
                  name="estimatedHours"
                  label="Estimated Hours"
                  type="number"
                  value={editedTicket.estimatedHours?.toString() ?? ""}
                  onChange={(e) => handleNumberChange(e, "estimatedHours")}
                  fullWidth
                  size="small"
                  InputProps={{
                    inputProps: { min: 0, step: 0.5 },
                  }}
                  disabled={disabled}
                />
                <TextField
                  name="remainingHours"
                  label="Remaining Hours"
                  type="number"
                  value={editedTicket.remainingHours?.toString() ?? ""}
                  onChange={(e) => handleNumberChange(e, "remainingHours")}
                  fullWidth
                  size="small"
                  InputProps={{
                    inputProps: { min: 0, step: 0.5 },
                  }}
                  disabled={disabled}
                />
                <TextField
                  name="completedHours"
                  label="Completed Hours"
                  type="number"
                  value={editedTicket.completedHours?.toString() ?? ""}
                  onChange={(e) => handleNumberChange(e, "completedHours")}
                  fullWidth
                  size="small"
                  InputProps={{
                    inputProps: { min: 0, step: 0.5 },
                  }}
                  disabled={disabled}
                />
              </TimeTrackingGrid>
            </Section>

            <Box
              display="flex"
              justifyContent="space-between"
              flexDirection={"column"}
              p={2}
              gap={1.5}
              mt="auto"
            >
              <Button
                fullWidth
                variant="outlined"
                onClick={onClose}
                sx={{ mr: 1 }}
                disabled={disabled}
              >
                Cancel
              </Button>
              <Button
                fullWidth
                variant="contained"
                color="primary"
                startIcon={<SaveIcon />}
                onClick={handleSave}
                disabled={disabled}
              >
                Save
              </Button>
              {onDelete && (
                <Button
                  fullWidth
                  variant="outlined"
                  color="error"
                  startIcon={<DeleteIcon />}
                  onClick={handleDelete}
                  disabled={disabled}
                >
                  Delete
                </Button>
              )}
            </Box>
          </RightPanel>
        </MainContent>
      </DialogContent>
    </StyledDialog>
  );
};

export default TicketEditDialog;
