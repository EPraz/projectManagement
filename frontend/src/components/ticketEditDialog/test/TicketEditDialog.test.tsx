import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import TicketEditDialog from "../TicketEditDialog";
import { MemoryRouter } from "react-router-dom";
import { Sprint, Ticket, User } from "../../../types";
import { TicketPriority, TicketType } from "../../../constants";

// Mocks de contexto
const mockAuth = {
  user: { email: "testuser@example.com" },
};
jest.mock("../../../context", () => ({
  useAuth: () => mockAuth,
  useSprint: () => ({
    sprint: { id: "sprint1" },
  }),
}));

// Datos de ejemplo
const sampleTicket: Ticket = {
  id: 1,
  title: "Initial Ticket Title",
  description: "Initial description",
  discussion: "Initial discussion",
  createdBy: "creator@example.com",
  updatedBy: "creator@example.com",
  createdAt: "2023-01-01T00:00:00.000Z",
  updatedAt: "2023-01-01T00:00:00.000Z",
  statusId: "status1",
  status: {
    id: "status1",
    name: "status1",
    position: 1,
  },
  order: 1,
  estimatedHours: 5,
  remainingHours: 3,
  completedHours: 2,
  priority: TicketPriority.HIGH,
  type: TicketType.BUG,
  tasks: [],
  assignedTo: "1",
  _count: null,
  pairProgrammingUsers: [],
};

const sampleStatuses = [
  { id: "status1", name: "OPEN", color: "#00ff00", position: 1 },
  { id: "status2", name: "BLOCKED", color: "#ff0000", position: 2 },
  { id: "status3", name: "CLOSED", color: "#0000ff", position: 3 },
];

const sampleUsers: User[] = [
  { id: "1", name: "Alice", email: "alice@example.com", projects: [] },
  { id: "2", name: "Bob", email: "bob@example.com", projects: [] },
];

const sampleSprints: Sprint[] = [
  {
    id: "sprint1",
    name: "Sprint 1",
    projectId: "123",
    createdAt: "asdf",
    updatedAt: "",
    sprintGoal: [],
    tickets: [],
    project: {
      id: "123",
      title: "Test Project",
      createdBy: "example",
      createdAt: "zxcv",
      updatedAt: "asdf",
      epics: [],
      tickets: [],
      users: [],
      sprints: [],
      ticketStatuses: [],
      taskStatuses: [],
      epicStatuses: [],
      featureStatuses: [],
      type: "hello",
      productManagers: [],
    },
  },
];

const sampleTickets = [sampleTicket];

describe("TicketEditDialog Component", () => {
  const onClose = jest.fn();
  const onSave = jest.fn();
  const onDelete = jest.fn();

  const renderComponent = (ticketProp = sampleTicket) =>
    render(
      <MemoryRouter>
        <TicketEditDialog
          open={true}
          ticket={ticketProp}
          onClose={onClose}
          onSave={onSave}
          onDelete={onDelete}
          statuses={sampleStatuses}
          users={sampleUsers}
          sprints={sampleSprints}
          features={[]}
          tickets={sampleTickets}
          disabled={false}
          projects={[]}
        />
      </MemoryRouter>
    );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("match snapshot", () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders dialog with initial ticket values", () => {
    renderComponent();
    const titleInput = screen.getByPlaceholderText(
      /ticket title/i
    ) as HTMLInputElement;
    expect(titleInput).toBeInTheDocument();
    expect(titleInput.value).toBe(sampleTicket.title);

    const descriptionInput = screen.getByPlaceholderText(
      /add a more detailed description/i
    );
    expect(descriptionInput).toHaveValue(sampleTicket.description);
  });

  it("updates title and calls onSave with updated data", async () => {
    renderComponent();
    const titleInput = screen.getByPlaceholderText(
      /ticket title/i
    ) as HTMLInputElement;
    fireEvent.change(titleInput, {
      target: { value: "Updated Ticket Title", name: "title" },
    });

    // Clic en Save
    const saveButton = screen.getByRole("button", { name: /save/i });
    fireEvent.click(saveButton);

    await waitFor(() => {
      expect(onSave).toHaveBeenCalledWith(
        expect.objectContaining({
          title: "Updated Ticket Title",
          updatedBy: "testuser@example.com",
          sprintId: "sprint1",
        })
      );
    });
    expect(onClose).toHaveBeenCalled();
  });

  it("calls onDelete and then onClose when Delete is clicked", async () => {
    renderComponent();
    const deleteButton = screen.getByRole("button", { name: /delete/i });
    fireEvent.click(deleteButton);
    await waitFor(() => {
      expect(onDelete).toHaveBeenCalledWith(sampleTicket);
    });
    expect(onClose).toHaveBeenCalled();
  });

  //   it("updates assignedTo when selecting a user from dropdown", async () => {
  //     renderComponent();
  //     // Buscar el select "Assigned To" por su label
  //     const assignedSelect = screen.getByLabelText(/assigned to/i);
  //     fireEvent.mouseDown(assignedSelect);
  //     // Seleccionamos la opci칩n "Bob"
  //     const bobOption = await screen.findByText("Bob");
  //     fireEvent.click(bobOption);
  //     // Ahora, para verificar, se espera que el select muestre "Bob"
  //     expect(assignedSelect).toHaveTextContent("Bob");
  //   });

  //   it("updates status when selecting a new status (if not blocked)", async () => {
  //     renderComponent();
  //     // Buscar el select "Status" por su label
  //     const statusSelect = screen.getByTestId("status-select");

  //     // const statusSelect = screen.getByLabelText(/status/i);
  //     fireEvent.mouseDown(statusSelect);
  //     // Seleccionamos la opci칩n "CLOSED"
  //     const closedOption = await screen.findByText(/closed/i);
  //     fireEvent.click(closedOption);
  //     expect(statusSelect).toHaveTextContent(/closed/i);
  //   });

  it("toggles blocking status when Chip is clicked", async () => {
    renderComponent();

    const blockedChip = screen.getByText(/blocked/i);

    fireEvent.click(blockedChip);

    const saveButton = screen.getByRole("button", { name: /save/i });
    fireEvent.click(saveButton);
    await waitFor(() => {
      // Como ejemplo: si se activa el bloqueo, se actualiza statusId con el id del status "BLOCKED" de sampleStatuses
      expect(onSave).toHaveBeenCalledWith(
        expect.objectContaining({
          isBlocked: true,
          statusId:
            sampleStatuses.find((s) => s.name === "BLOCKED")?.id ||
            expect.any(String),
        })
      );
    });
  });

  it("updates number fields when changed", () => {
    renderComponent();
    const estimatedInput = screen.getByLabelText(
      /estimated hours/i
    ) as HTMLInputElement;
    fireEvent.change(estimatedInput, {
      target: { value: "8", name: "estimatedHours" },
    });
    expect(estimatedInput.value).toBe("8");
  });

  //   it("renders additional tabs and changes content based on activeTab", () => {
  //     renderComponent();
  //     // Verifica que se renderice la pesta침a "Details" inicialmente
  //     expect(screen.getByText(/ticket information/i)).toBeInTheDocument();
  //     // Simula cambiar a la pesta침a "Planning"
  //     const planningTab = screen.getByRole("tab", { name: /planning/i });
  //     fireEvent.click(planningTab);
  //     expect(screen.getByText(/sprint/i)).toBeInTheDocument();
  //     // Cambia a "Additional Info"
  //     const additionalTab = screen.getByRole("tab", { name: /additional info/i });
  //     fireEvent.click(additionalTab);
  //     expect(screen.getByText(/acceptance criteria/i)).toBeInTheDocument();
  //   });
});
