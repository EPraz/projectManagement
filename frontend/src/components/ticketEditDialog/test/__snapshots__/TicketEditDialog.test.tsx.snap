// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketEditDialog Component match snapshot 1`] = `<DocumentFragment />`;
